tfhe/src/integer/server_key/radix_parallel/ilog2.rs

```diff
@@ -0,0 +1,734 @@
+    //==============================================================================================
+
+    /// Returns the number of trailing zeros in the binary representation of `ct`
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use tfhe::integer::gen_keys_radix;
+    /// use tfhe::shortint::parameters::PARAM_MESSAGE_2_CARRY_2;
+    ///
+    /// // Generate the client key and the server key:
+    /// let num_blocks = 4;
+    /// let (cks, sks) = gen_keys_radix(PARAM_MESSAGE_2_CARRY_2, num_blocks);
+    ///
+    /// let msg = -4i8;
+    ///
+    /// let ct1 = cks.encrypt_signed(msg);
+    ///
+    /// let n = sks.trailing_zeros(&ct1);
+    ///
+    /// // Decrypt:
+    /// let n: u32 = cks.decrypt(&n);
+    /// assert_eq!(n, msg.trailing_zeros());
+    /// ```
+    pub fn trailing_zeros<T>(&self, ct: &T) -> RadixCiphertext
+    where
+        T: IntegerRadixCiphertext,
+    {
+        let mut tmp;
+        let ct = if ct.block_carries_are_empty() {
+            ct
+        } else {
+            tmp = ct.clone();
+            self.full_propagate_parallelized(&mut tmp);
+            &tmp
+        };
+        self.unchecked_trailing_zeros(ct)
+    }
+
+    /// Returns the number of trailing ones in the binary representation of `ct`
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use tfhe::integer::gen_keys_radix;
+    /// use tfhe::shortint::parameters::PARAM_MESSAGE_2_CARRY_2;
+    ///
+    /// // Generate the client key and the server key:
+    /// let num_blocks = 4;
+    /// let (cks, sks) = gen_keys_radix(PARAM_MESSAGE_2_CARRY_2, num_blocks);
+    ///
+    /// let msg = -4i8;
+    ///
+    /// let ct1 = cks.encrypt_signed(msg);
+    ///
+    /// let n = sks.trailing_ones(&ct1);
+    ///
+    /// // Decrypt:
+    /// let n: u32 = cks.decrypt(&n);
+    /// assert_eq!(n, msg.trailing_ones());
+    /// ```
+    pub fn trailing_ones<T>(&self, ct: &T) -> RadixCiphertext
+    where
+        T: IntegerRadixCiphertext,
+    {
+        let mut tmp;
+        let ct = if ct.block_carries_are_empty() {
+            ct
+        } else {
+            tmp = ct.clone();
+            self.full_propagate_parallelized(&mut tmp);
+            &tmp
+        };
+        self.unchecked_trailing_ones(ct)
+    }
+
+    /// Returns the number of leading zeros in the binary representation of `ct`
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use tfhe::integer::gen_keys_radix;
+    /// use tfhe::shortint::parameters::PARAM_MESSAGE_2_CARRY_2;
+    ///
+    /// // Generate the client key and the server key:
+    /// let num_blocks = 4;
+    /// let (cks, sks) = gen_keys_radix(PARAM_MESSAGE_2_CARRY_2, num_blocks);
+    ///
+    /// let msg = -4i8;
+    ///
+    /// let ct1 = cks.encrypt_signed(msg);
+    ///
+    /// let n = sks.leading_zeros(&ct1);
+    ///
+    /// // Decrypt:
+    /// let n: u32 = cks.decrypt(&n);
+    /// assert_eq!(n, msg.leading_zeros());
+    /// ```
+    pub fn leading_zeros<T>(&self, ct: &T) -> RadixCiphertext
```

Comment:
we should probably have an ilog2 function for completeness and perhaps a ceil_ilog2

Reply:
Its meant for the next PR